service.cc

```diff
@@ -0,0 +1,348 @@
+  std::string user_exists = backend_client_.SendGetRequest(request->username());
+
+  if(user_exists.empty()) {
+    return grpc::Status(grpc::StatusCode::NOT_FOUND, "User not found");
+  }
+
+  // check for parent id's existance
+  std::string to_follow_exists = backend_client_.SendGetRequest(request->to_follow());
+  if(to_follow_exists.empty()) {
+    return grpc::Status(grpc::StatusCode::NOT_FOUND, "Requested user to follow not found");
+  }
+
+  // Update user with new follower (to_follow), send to backend to update
+  chirp::User user;
+  user.ParseFromString(user_exists);
+  std::vector<std::string> currently_following;
+  std::copy(user.following().begin(), user.following().end(), std::back_inserter(currently_following));
+
+  // don't follow someone more than once
+  for(std::string following : currently_following) {
+    if(following == request->to_follow()) {
+      return grpc::Status(grpc::StatusCode::ALREADY_EXISTS, "Already following");
+    }
+  }
+  user.add_following(request->to_follow()); // add as follower
+  user.SerializeToString(&user_exists); 
+  backend_client_.SendPutRequest(request->username(), user_exists);
+
+  return grpc::Status::OK;
+}
+
+void Service::ThreadDFS(chirp::ReadReply *reply, std::string chirp_id, int depth)
+{
+  // Get chirp from backend and get IDs of all children (replies)
+  chirp::Chirp chirp;
+  std::string chirp_serialized = backend_client_.SendGetRequest(chirp_id);
+  chirp.ParseFromString(chirp_serialized);
+
+  std::vector<std::string> children_chirps_serialized; 
+  std::copy(chirp.child_ids().begin(), chirp.child_ids().end(), std::back_inserter(children_chirps_serialized));
+
+  // if chirp hasn't been visited, add to reply and 
+  // set depth of reply - used for nested printing in client
+  if(chirp.depth() == 0) {
+      chirp.set_depth(depth);
+      chirp::Chirp *new_chirp = reply->add_chirps();
+      *new_chirp = chirp;
+  }
+  // chirp has no children recurse up -- base case
+  if(children_chirps_serialized.empty()) {
+    return;
+  }
+  
+  // loop thru child chirps
+  for(std::string child_chirp : children_chirps_serialized) {
+    ThreadDFS(reply, child_chirp, chirp.depth()+1); 
+  }
+}
+
+grpc::Status Service::read(grpc::ServerContext *context, const chirp::ReadRequest *request, chirp::ReadReply *reply) {
+  // Check if chirp ID exists in key value store
+  chirp::Chirp chirp;
+  std::string chirp_serialized = backend_client_.SendGetRequest(request->chirp_id());
+
+  if(chirp_serialized.empty()) {
+    return grpc::Status(grpc::StatusCode::NOT_FOUND, "Chirp not found");
+  }
+
+  // Chirp was found, perform dfs to collect entire thread
+  std::vector<chirp::Chirp> *chirp_thread;
+  ThreadDFS(reply, request->chirp_id(), 1);
+  return grpc::Status::OK;
+}
+
+grpc::Status Service::monitor(grpc::ServerContext *context, const chirp::MonitorRequest *request, grpc::ServerWriter<chirp::MonitorReply> *stream) {
+  chirp::MonitorReply reply;
```

Comment:
This function is a bit too long -- is there a way to break it up with a few helpers?

Reply:
yes I will split up